### PR TITLE
Issue #3012 refactored HttpCompliance

### PR DESCRIPTION
**Pull request by gregw**
#3012 HttpCompliance
This is an alternative to #3029, currently just refactoring `HttpCompliance`

While I understand there are concepts of `SpecReferences` and `SpecViolations` that are conceptually independent of the optional behaviours in `HttpCompliance`, in practise there just isn't enough code that needs those concepts to justify the work to establish it. There are perhaps a few throws of `BadMessageException` in `HttpParser` that could be better with a spec reference, but generally our exceptions do not reference specs and that is a big task to always do that.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

**Comment by gregw** [comment]
@joakime The multipart compliance mechanism was removed as part of #3398 / #1676 ?  Was that intentional?   In that case, I think this PR is complete!

**Comment by joakime** [comment]
I'll review the current state of this tomorrow.

**Comment by gregw** [comment]
@joakime I'm inclined to think multipart compliance can be removed, since if somebody really wants to the old behaviour, we could create a filter to wrap the request and use the old parser.  Ie there is a way to get the old behaviour without switching in the container.  

@lachlan-roberts if we keep the MP compliance removed... are there any other simplifications that can be done to that code - ie abstractions removed that hid the two impls?

**Comment by joakime** [comment]
There are lot of test failures on this PR.

**Comment by joakime** [comment]
Missing work on CookieCompliance and CookieCutter violations reporting (and testing)

* COMMA_NOT_VALID_OCTET / https://tools.ietf.org/html/rfc6265 / Reserved $Names no longer use '$' prefix
* RESERVED_NAMES_NOT_DOLLAR_PREFIXED / https://tools.ietf.org/html/rfc6265#page-9 / Comma ',' not valid as cookie-octet or separator

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java [hunk 1]
```diff
@@ -19,99 +19,127 @@
...
-    /** Jetty's current RFC7230 support, which incorrectly excludes  {@link HttpComplianceSection#METHOD_CASE_SENSITIVE} */
-    RFC7230_LEGACY(sectionsBySpec("RFC7230,-METHOD_CASE_SENSITIVE")),
 
-    /** The RFC7230 support mode */
-    RFC7230(sectionsBySpec("RFC7230"));
+    // These are compliance violations, which may optionally be allowed by the compliance mode, which mean that
+    // the relevant section of the RFC is not strictly adhered to.
+    public enum Violation
+    {
+        CASE_SENSITIVE_FIELD_NAME("https://tools.ietf.org/html/rfc7230#section-3.2", "Field name is case-insensitive"),
```
Comment (joakime) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: Would really like to see `ALLOW_*` and `REQUIRE_*` prefixes on these.

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java [hunk 2]
```diff
@@ -2295,7 +2310,13 @@ public int getHeaderCacheSize()
...
 
         @Override
-        public void onComplianceViolation(HttpCompliance compliance, HttpComplianceSection violation, String reason)
+        public boolean isHeaderCacheCaseSensitive()
+        {
+            return _headerCacheCaseSensitive;
+        }
+
+        @Override
+        public void onComplianceViolation(HttpCompliance compliance, HttpCompliance.Violation violation, String reason)
```
Comment (joakime) [inline jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: This isn't limited to HttpCompliance. CookieCompliance can, and should, produce violations as well.
There's also violations produced and reported during Parameter decoding (query / form-data / multipart)

File: jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java [hunk 3]
```diff
@@ -765,9 +765,9 @@ public boolean onEarlyEOF()
     public void onBadMessage(BadMessageException failure)
     {
         int status = failure.getCode();
-        String reason = failure.getReason();
+        String message = failure.getReason();
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java]: I think you should revert the changes in this file, these seem to be related to a different issue. (such as #3225)

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java [hunk 4]
```diff
@@ -19,99 +19,127 @@
...
+            Violation.NO_COLON_AFTER_FIELD_NAME,
+            Violation.TRANSFER_ENCODING_WITH_CONTENT_LENGTH,
+            Violation.MULTIPLE_CONTENT_LENGTHS);
+    public final static HttpCompliance RFC7230_LEGACY = RFC7230.with("RFC7230_LEGACY", Violation.CASE_INSENSITIVE_METHOD);
 
     public static final String VIOLATIONS_ATTR = "org.eclipse.jetty.http.compliance.violations";
 
     private static final Logger LOG = Log.getLogger(HttpParser.class);
-    private static EnumSet<HttpComplianceSection> sectionsByProperty(String property)
+    private static EnumSet<Violation> violationByProperty(String property)
```
Comment (joakime) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: This method isn't used?

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java [hunk 5]
```diff
@@ -1945,15 +1939,15 @@ public default void badMessage(BadMessageException failure)
...
          * @return true if handling parsing should return
          */
-        public boolean startResponse(HttpVersion version, int status, String reason);
+        boolean startResponse(HttpVersion version, int status, String reason);
     }
 
     /* ------------------------------------------------------------------------------- */
     /* ------------------------------------------------------------------------------- */
     /* ------------------------------------------------------------------------------- */
     public interface ComplianceHandler extends HttpHandler
```
Comment (joakime) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: Move this out into a top level interface, it's no longer specific to HttpParser.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java [hunk 6]
```diff
@@ -1945,15 +1939,15 @@ public default void badMessage(BadMessageException failure)
...
     {
-        public default void onComplianceViolation(HttpCompliance compliance, HttpComplianceSection violation, String details)
+        default void onComplianceViolation(HttpCompliance compliance, HttpCompliance.Violation violation, String details)
```
Comment (joakime) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: This is not specific to HttpCompliance, violation reports can come from (currently) HttpParser, CookieCutter, Request.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java (hunk 1 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: I've gone back and forth on this naming, as I agree that it is kind of attractive to have ALLOW_* and REQUIRE_* in these names.  However, what that does is allow both negative and positive Violations to exist in the enum and would prevent the usage of nice syntax like: `if (CASE_SENSITIVE_FIELD_NAME.isAllowedBy(_compliance))`.      So in the end I think it is best to ensure that all the Violations are actually violations of the RFC/Spec, so that their inclusion in a compliance mode allows us to ask   " if this Violation is allowed by this compliance mode?"

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java (hunk 2 above)
Comment (gregw) [reply jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: OK, so much like your PR we should have `HttpCompliance.Violation` implement a common `Violation` interface which is used for reporting.      Perhaps the `Violation` interface could use a `SpecReference` interface?... or maybe we do that later when we have a usage of SpecReference separate from a Violation.

File: jetty-fcgi/fcgi-client/src/main/java/org/eclipse/jetty/fcgi/parser/ResponseContentParser.java [hunk 7]
```diff
@@ -157,6 +157,13 @@ public int getHeaderCacheSize()
             return 4096;
         }
 
+        @Override
+        public boolean isHeaderCacheCaseSensitive()
+        {
+            // TODO get from configuration
```
[the remark this answers is not in the document]
Comment (gregw) [reply jetty-fcgi/fcgi-client/src/main/java/org/eclipse/jetty/fcgi/parser/ResponseContentParser.java]: This TODO is just echoing the same TODO in the related `getHeaderCacheSize()`.   So it should be fixed, but is outside of the scope of the PR to work out how to configure the header cache in all these circumstances.  @sbordet  did you have a plan for this?

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java [hunk 8]
```diff
@@ -1223,24 +1216,23 @@ else if (_endOfContent == EndOfContent.UNKNOWN_CONTENT)
                                     String n = cached_field.getName();
                                     String v = cached_field.getValue();
 
-                                    if (!_compliances.contains(HttpComplianceSection.FIELD_NAME_CASE_INSENSITIVE))
+                                    if (CASE_SENSITIVE_FIELD_NAME.isAllowedBy(_compliance))
```
[the remark this answers is not in the document]
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: thanks. Using it more extensively now.

File: jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelState.java [hunk 9]
```diff
@@ -740,13 +740,13 @@ protected void onError(Throwable th)
         final Request baseRequest = _channel.getRequest();
         
         int code=HttpStatus.INTERNAL_SERVER_ERROR_500;
-        String reason=null;
+        String message=null;
```
[the remark this answers is not in the document]
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelState.java]: not exactly, because we were treating some of the violation descriptions as reasons, which is not correct.  They should be treated as messages and this renaming makes that clear.  So a little out of scope... but near enough that I'm OK with including it.

File: tests/jetty-http-tools/src/main/java/org/eclipse/jetty/http/tools/HttpTester.java [hunk 10]
```diff
@@ -512,6 +512,12 @@ public int getHeaderCacheSize()
         {
             return 0;
         }
+
+        @Override
+        public boolean isHeaderCacheCaseSensitive()
+        {
+            return false;
```
[the remark this answers is not in the document]
Comment (gregw) [reply tests/jetty-http-tools/src/main/java/org/eclipse/jetty/http/tools/HttpTester.java]: currently yes.  Can be made configurable if there is a use-case

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java (hunk 2 above)
Comment (gregw) [reply jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: done

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java (hunk 4 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: will remove in next commit

**Comment by gregw** [comment]
@joakime I've implemented CookieCompliance.  But do you have a preference or idea for how to wire up ComplianceViolation.Listener for cookies?

**Comment by joakime** [comment]
@gregw i added the cookie compliance listener parts, fixed some javadoc.
Merge it if you are ok with this version.

**Comment by joakime** [comment]
@gregw don't forget to delete the branch (see "Delete Branch") button at the bottom of this PR.
